swarm/storage/feed/lookup/lookup_test.go

```diff
@@ -237,99 +285,132 @@ func TestBadHintNextToUpdate(t *testing.T) {
+	// ### 1.- Test all algorithms
+	for _, algo := range algorithms {
+		t.Run(algo.Name, func(t *testing.T) {
+
+			// ### 2.1.- Test a simple cancel of an always blocking read function
+			readFunc := func(ctx context.Context, epoch lookup.Epoch, now uint64) (interface{}, error) {
+				<-ctx.Done()
+				return nil, ctx.Err()
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			errc := make(chan error)
+
+			go func() {
+				_, err := algo.Lookup(ctx, 1200000000, lookup.NoClue, readFunc)
+				errc <- err
+			}()
+
+			cancel() //actually cancel the lookup
+
+			if err := <-errc; err != context.Canceled {
+				t.Fatalf("Expected lookup to return a context canceled error, got %v", err)
+			}
+
+			// ### 2.2.- Test context cancellation during hint lookup:
+			ctx, cancel = context.WithCancel(context.Background())
+			errc = make(chan error)
+			someHint := lookup.Epoch{
+				Level: 25,
+				Time:  300,
+			}
+			// put up a read function that gets canceled only on hint lookup
+			readFunc = func(ctx context.Context, epoch lookup.Epoch, now uint64) (interface{}, error) {
+				if epoch == someHint {
+					go cancel()
+					<-ctx.Done()
+					return nil, ctx.Err()
+				}
+				return nil, nil
+			}
+
+			go func() {
+				_, err := algo.Lookup(ctx, 301, someHint, readFunc)
+				errc <- err
+			}()
+
+			if err := <-errc; err != context.Canceled {
+				t.Fatalf("Expected lookup to return a context canceled error, got %v", err)
+			}
+		})
 	}
 
 }
 
+// TestLookupFail makes sure the lookup function fails on a timely manner
+// when there are no updates at all
 func TestLookupFail(t *testing.T) {
+	// ### 1.- Initialize stopwatch time sim
+	stopwatch := NewStopwatch(50 * time.Millisecond)
+	lookup.TimeAfter = stopwatch.TimeAfter()
+	defer stopwatch.Stop()
 
-	store := make(Store)
-	readCount := 0
+	// ### 2.- Setup mock storage, without adding updates
+	// don't write anything and try to look up.
+	// we're testing we don't get stuck in a loop and that the lookup
+	// function converges in a timely fashion
+	store := NewStore(DefaultStoreConfig)
+	readFunc := store.MakeReadFunc()
 
-	readFunc := makeReadFunc(store, &readCount)
 	now := uint64(1533903729)
 
-	// don't write anything and try to look up.
-	// we're testing we don't get stuck in a loop
+	// ### 3.- Test all algorithms
+	for _, algo := range algorithms {
+		t.Run(algo.Name, func(t *testing.T) {
+			store.Reset()
 
-	value, err := lookup.Lookup(context.Background(), now, lookup.NoClue, readFunc)
-	if err != nil {
-		t.Fatal(err)
-	}
-	if value != nil {
-		t.Fatal("Expected value to be nil, since the update should've failed")
-	}
+			stopwatch.Measure(func() {
+				value, err := algo.Lookup(context.Background(), now, lookup.NoClue, readFunc)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if value != nil {
+					t.Fatal("Expected value to be nil, since the update should've failed")
+				}
+			})
 
-	expectedReads := now/(1<<lookup.HighestLevel) + 1
```

Comment:
The new algorithm would fail this test since it does not necessarily read fewer times when a hint is given. It takes less time, but does not make fewer reads